File: apiRestful/src/main/java/Service/StudentService.java
package Service;

import Database.DataRepository;
import Model.Student;
import com.google.gson.Gson;
import spark.Request;
import java.util.List;
import static spark.Spark.*;

/**
 * @author carloscastillo
 * @author jhandrylopez
 * author gersonsantos
 */

public class StudentService {
    public static void main(String[] args){

        get("/students", (req, res) -> {
           Gson gson = new Gson();
           res.type("application/json");
           return gson.toJson(DataRepository.getStudents());
        });


        get("/students/filter", (req, res) -> {
            var key = req.queryParams().stream().toList().get(0);
            Gson gson = new Gson();
            res.type("application/json");
            List<Student> students = validateFilter(key, req);

            if(students != null && !students.isEmpty()){
                res.status(200);
            }else{
                return halt(404);
            }

            return gson.toJson(students);
        });


        get("/student/:pos", (req, res) -> {
            var pos = Integer.parseInt(req.params(":pos"));
            Gson gson = new Gson();
            res.type("application/json");
            Student student = DataRepository.searchStudentByPos(pos);
            if(student != null){
                res.status(200);
            }else{
                halt(404);
            }
            return gson.toJson(student);
        });


        post("/student", (req, res) -> {
            var dataStudent = req.body().toString();
            int position = DataRepository.addStudent(dataStudent);
            res.type("application/json");

            if(position > 0){
                res.status(201);
            }else{
                halt(400);
            }

            return "Location: https://88e4-190-110-60-131.ngrok.io/student/" + position;
        });


        put("/student/:pos", (req, res) -> {
            var pos = Integer.parseInt(req.params(":pos"));
            var dataStudent = req.body().toString();
            int position = DataRepository.putStudent(pos, dataStudent);
            res.type("application/json");

            if(position > 0){
                res.status(204);
            }else{
                halt(409);
            }
            return halt(204);
        });


        delete("/student/:pos", (req, res) -> {
            try {
                var pos = Integer.parseInt(req.params(":pos"));
                DataRepository.deleteStudentByPos(pos);
                res.type("application/json");
                res.status(204);
            }catch (Exception e){
                halt(400);
            }

            return halt(204);
        });

    }

    private static List<Student> validateFilter(String key, Request req){
        try {
            var value = req.queryParams(key);
            List<Student> students = null;
            if(key.equals("name")){
                students = DataRepository.searchStudentByName(value);
            }else{
                if(key.equals("lastname")){
                    students = DataRepository.searchStudentByLastname(value);
                }else{
                    if(key.equals("age")){
                        students = DataRepository.searchStudentByAge(Integer.parseInt(value));
                    }
                }
            }
            return students;
        }catch(Exception e){
            return null;
        }
    }
}

File: python-flask/apiRestful/api/service/student_service.py
from api.enconde import EmployeeEncoder
from api.database.data_repository import DataRepository
from flask import abort, request
from flask_negotiate import consumes, produces
from api.service import blueprint
import json

data_repository = DataRepository()


@blueprint.route('/')
def index():
    return "Bienvenido"

@blueprint.route('/students', methods=['GET'])

@produces('application/json')
def get_students():
    students = data_repository.get_all_student()
    if students:
        return json.dumps(students, cls=EmployeeEncoder), 200

    abort(404)


@blueprint.route('/students/filter', methods=['GET'])
@produces('application/json')
def get_student_by_filter():
    students = validate_filter(request.args.to_dict())
    if students:
        return json.dumps(students, cls=EmployeeEncoder), 200
    abort(404)


@blueprint.route('/student/<int:pos>', methods=['GET'])
@produces('application/json')
def get_student_by_post(pos):
    student = data_repository.search_student_by_pos(pos)
    if student:
        return json.dumps(student, cls=EmployeeEncoder), 200
    abort(404)


@blueprint.route('/student', methods=['POST'])
@consumes('application/json')
def post_student():
    try:
        pos = data_repository.add_student(request.get_json())
        return "Location: {}/{}".format(request.url, pos), 201
    except:
        abort(400)


@blueprint.route('/student/<int:pos>', methods=['PUT'])
@consumes('application/json')
def put_student_by_pos(pos):
    try:
        data_repository.update_student(pos, request.get_json())
        return "", 204
    except:
        abort(409)


@blueprint.route('/student/<int:pos>', methods=['DELETE'])
@produces('application/json')
def delete_student_by_pos(pos):
    try:
        data_repository.delete_student(pos)
        return "", 204
    except:
        abort(400)


def validate_filter(query_params):
    key = list(query_params.keys())[0]
    value = query_params[key]
    students = []

    if key == 'name':
        students = data_repository.search_student_by_name(value)

    elif key == 'lastname':
        students = data_repository.search_student_by_lastname(value)

    elif key == 'age':
        students = data_repository.search_student_by_age(int(value))

    return students
